Propagate exceptions fully when using proc_lib

essen opened the pull request:
This is a small and heavy change. I can do any documentation changes after the PR is approved.
## The patch

This makes proc_lib behaves like a normal process as far
as exceptions are concerned.

Before this commit, the following difference could be
observed:

```
6> spawn_link(fun() -> ssl:send(a,b) end).
<0.43.0>
7> flush().
Shell got {'EXIT',<0.43.0>,
           {function_clause,
            [{ssl,send,[a,b],[{file,"..."},{line,275}]}]}}
ok

8> proc_lib:spawn_link(fun() -> ssl:send(a,b) end).
<0.46.0>
9> flush().
Shell got {'EXIT',<0.46.0>,function_clause}
```

After this commit, we get the following instead:

```
3> flush().
Shell got {'EXIT',<0.61.0>,
           {function_clause,
            [{ssl,send,[a,b],[{file,"..."},{line,275}]},
             {proc_lib,init_p,3,[{file,"..."},{line,232}]}]}}
```

The stacktrace will show minor differences of course
but the form is now the same as without proc_lib.
## Rationale
- We now have a single form regardless of how the process
  was started
- We can use the stacktrace to programmatically alter behavior
  (for example an HTTP server identifying problems in input
  decoding to send back a generic 400, or a 500 otherwise)
- We can access the stacktrace to print it somewhere (for
  example an HTTP server could send it back to the client
  when a debug mode is enabled)
## Breaking changes
- The 'EXIT' message reason format changes due to the addition of the stacktrace

essen commented:
Making some clarifications about the beginners comment. When teaching people how to write a gen_server (as part of learning OTP), the error reports are all gone as soon as they add proc_lib.

josevalim commented on lib/stdlib/src/proc_lib.erl:
```diff
@@ -265,12 +265,12 @@ exit_p(Class, Reason) ->
 	    MFA = {M,F,make_dummy_args(A, [])},
 	    crash_report(Class, Reason, MFA),
-	    exit(Reason);
+	    erlang:raise(Class, Reason, erlang:get_stacktrace());
```
You will want to pass `erlang:get_stacktrace()` as an argument to `exit_p/2` instead of calculating it here. The reason is that if any of the code between calling `exit_p/2` and calling `erlang:get_stacktrace()` uses `catch` or `try`, it can potentially change the value of `erlang:get_stacktrace()`. To avoid such bugs being introduced, I typically call `erlang:get_stacktrace()` as the first thing inside my catch clauses.

essen replied:
Done. I didn't think it was really necessary considering the module but better safe than sorry.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

essen commented:
After a week of soul searching I think I will make it raise an exit exception every time (same as currently but with stacktrace propagated) and instead amend the proc_lib manual to indicate that proc_lib processes hide normal error reports and produce crash reports instead. There's already mention of crash reports, but it's not clear as to what happens to error reports.

essen commented:
I have updated the PR with the previously mentioned changes. Exceptions are now all exits (like before, just with the stacktrace propagated), and I have added a paragraph in the docs to explain why there's no error reports.

I had a hard time producing the doc paragraph to be even remotely readable so if there's any comments I will happily address them.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

bjorng commented:
Can you please rebase this branch on the latest master?

We want to test it in our daily builds. If everything goes well, we intend to merge it to master for inclusion in OTP 20.

essen commented:
@bjorng Rebased to master and pushed. Sorry it took so long, had some email problems with Github when you replied and so didn't see.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

bjorng commented:
`proc_lib_SUITE:stop/1` fails like this:

```
=== Location: [{proc_lib_SUITE,stop,460},
              {test_server,ts_tc,1529},
              {test_server,run_test_case_eval1,1045},
              {test_server,run_test_case_eval,977}]
=== Reason: no match of right hand side value {'EXIT',{badmatch,2}}
  in function  proc_lib_SUITE:stop/1 (proc_lib_SUITE.erl, line 460)
  in call from test_server:ts_tc/3 (test_server.erl, line 1529)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1045)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 977)

```

essen commented:
Thanks. Weird, I thought I fixed that. I'll look tonight.

essen commented:
Pushed the fix. I think my previous change did it. I probably messed up when trying to re-run tests.

Also added a test case with exits caused by error, exit and throw.

bjorng commented:
Thanks! The test case now succeeds.

One thing: The indentation for the new `stacktrace/1` function is different from the indentation in the rest of the file. Please correct the indentation for the new function.

essen commented:
Done. Hopefully didn't mess up there. Mixed tabs and spaces is hard.

bjorng commented:
Thanks! Looks good!

bjorng commented:
Regarding indentation, we have recently changed the guidelines about indentation. We now recommend indenting with only spaces, but we still accept mixed tabs and spaces. (Indentation is still 4 steps.)

The guideline about indentation is at the end of this page:

https://github.com/erlang/otp/wiki/Contribution-Guidelines

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

bjorng commented:
There are now three failed test cases in `gen_statem_SUITE`.

essen commented:
Done.

bjorng commented:
Thanks!

I everything goes well, I plan to merge this pull request on Wednesday.
